qiskit/circuit/parametertable.py

```diff
@@ -45,9 +49,37 @@ def __setitem__(self, parameter, instr_params):
             assert isinstance(instruction, Instruction)
             assert isinstance(param_index, int)
         self._table[parameter] = instr_params
+        self._keys.add(parameter)
+        self._names.add(parameter.name)
+
+    def get_keys(self):
+        """Return a set of all keys in the parameter table
+
+        Returns:
+            set: A set of all the keys in the parameter table
+        """
+        if len(self._keys) != len(self._table):
```

Comment:
There's no foreseeable case where this should happen, right? Would it be better to raise in this case (so we can identify the case we're missing, and keep consistent performance)?

Reply:
I actually hit this case in the unit tests, which is why I added this here: https://github.com/Qiskit/qiskit-terra/pull/4232/commits/995877ef6deb354373c98e6d77d5e1f22ecdea40 from what I could tell it was from tests that ran `QuantumCircuit.copy()` somehow the table would be populated but the keys set would be empty in the new object resulting from the copy. I was at a loss for how that could happen so I just added this fallback. I agree it would be better to avoid this, but I didn't have a better idea.

Reply:
The issue here is likely that when we copy a `QuantumCircuit`, we directly populate `ParameterTable._table`, bypassing `__setitem__`: https://github.com/Qiskit/qiskit-terra/blob/b84ace8/qiskit/circuit/quantumcircuit.py#L1223 . Probably this should be updated to pass the mapped `_table` via `ParameterTable__init__`

Reply:
ok, I'll update this in a sec

Reply:
Fixed in: https://github.com/Qiskit/qiskit-terra/pull/4232/commits/eb97e293f22cb63f873a83ba2df1830754d24d2c